python/pyiceberg/avro/reader.py

```diff
@@ -153,8 +154,8 @@ def skip(self, decoder: BinaryDecoder) -> None:
 
 
 class DateReader(Reader):
-    def read(self, decoder: BinaryDecoder) -> int:
-        return decoder.read_int()
+    def read(self, decoder: BinaryDecoder) -> date:
```

Comment:
I don't think this should produce a `date`. Our internal representation for date/time objects are:
* `int` days from epoch for `date`
* `int` (`long`) micros from epoch for `timestamp` and `timestamptz`
* `int` micros from midnight for `time`

We do not want to use native date/time representations internally because they require a lot of extra logic to compare and work with in other ways.

Reply:
Yes, I reverted this, and also fixed this for the `datetime`, that would actually break. I've also added some integration tests to cover this.